### PR TITLE
[IIIF-887] Add manifest component ID minter

### PR DESCRIPTION
**Pull request by ksclarke**
* Add ID minter for manifest components + a MinterFactory so different implementations can be swapped in (the tests use different fake minters to confirm this works).
* Allow manifest components that take an ID relative to the manifest to take a minter in their constructors.
* Update `paintWith()` and `supplementWith()` to take a minter.
* Change protected `paintWith()` and `supplementWith()` methods in AbstractCanvas to be `paint(CanvasResource [...])` and `supplement(CanvasResource [...])`; public methods in the canvas resources that extend AbstractCanvas are still `paintWith()` and `supplementWith()` to conform to the CanvasResource interface -- this allows passing in the canvas resource, behind the scenes, to give to the minter.
* Reduce the number of `paintWith()` and `supplementWith()` protected methods in the AbstractCanvas and have the corresponding overriding methods do more of the work of creating MediaFragmentSelector(s), etc.
* Change MediaFragmentSelector's `getValue()` to `toString()` to match other existing selectors' use of `toString()`.
* Change AbstractCanvas to CanvasResource in places where a canvas resource that extends AbstractCanvas is passed in to a method or constructor.
* Add missing Javadocs (that Checkstyle didn't catch because the methods were protected, but whose public extending methods referenced by @ Override(s)).
* Change Annotation's `getTarget()` to return a URI instead of an Object that needs casting and add `getSpecificResourceTarget()` and `hasSpecificResourceTarget()` for targets that are specific resources; add `getTargetObject()` that returns an Object for Jackson's serialization.
* Update some minor formatting due to changes in upstream Checkstyle rules.
* Add a TestUtils method that ignores IDs in JsonObjects to simplify tests where structure, not ID, is what's being tested.
* Add SupplementaryAnnotations on Range (not really related to this ticket, but is something that slipped by without getting its own ticket).
* Move private methods down in the SpecificResource class and make `setSelector()` private so it can't be changed after the SpecificResource has been created (which would break some uses).

A little note on the default minter implementation: it creates a maximum number of 1500625 IDs/NOIDs to be used in the construction of sub-manifest IDs (so for canvases, annotations, annotation pages, etc.) The IDs/NOIDs are unique to the manifest, not globally unique (until they're combined with the manifest ID). We've talked about having the Annotation IDs be independent from the Manifest in the future (with ARKs, perhaps), so they don't try to maintain the path within the particular Manifest as a part of their ID. We want to give the appearance of randomness of the NOIDs (not just be sequential IDs like 001, 002, 003...) so that inserting other canvases, etc., later doesn't break a perceived ordering of IDs. We chose four character alphanumeric strings, excluding the lower case 'L' (because it looks too much like a "1", which could confuse people). To give the appearance of randomness, the NOIDs are created and shuffled for the minter. Then, a particular use of the minter (for a given manifest) gets an iterator for the list that iterates through the list in a "random" order (e.g., starting at X and skipping Y numbers of NOIDs for each cycle through the list until all the NOIDs have been used). The minter creates IDs from the NOIDs for each type of resource it knows how to handle (e.g., Range, AnnotationPage, Canvas, etc.) This is a first pass at this.

### CODE REVIEW

**Comment by markmatney** [comment]
Looks great, just a few minor things.

File: src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java [hunk 1]
```diff
@@ -235,79 +265,86 @@ public AccompanyingCanvas setWidthHeight(final int aWidth, final int aHeight) {
...
+        return (AccompanyingCanvas) super.paint(this, aMinter, new MediaFragmentSelector(aCanvasRegion),
+                aContentArray);
     }
 
     @Override
-    public final AccompanyingCanvas paintWith(final MediaFragmentSelector aCanvasRegion,
-            final List<ContentResource> aContentList) throws IllegalArgumentException {
-        return (AccompanyingCanvas) super.paintWith(aCanvasRegion, aContentList);
+    public final AccompanyingCanvas paintWith(final Minter aMinter, final MediaFragmentSelector aCanvasRegion,
+            final List<ContentResource> aContentList) throws ContentOutOfBoundsException,
```
[the remark this answers is not in the document]
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java:296]: This has that on line 297

File: src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java [hunk 2]
```diff
@@ -235,79 +265,86 @@ public AccompanyingCanvas setWidthHeight(final int aWidth, final int aHeight) {
...
+        return (AccompanyingCanvas) super.paint(this, aMinter, aCanvasRegion, aContentList.toArray(
+                new ContentResource[] {}));
     }
 
     @Override
-    public final AccompanyingCanvas paintWith(final String aCanvasRegion, final List<ContentResource> aContentList)
-            throws IllegalArgumentException {
-        return (AccompanyingCanvas) super.paintWith(aCanvasRegion, aContentList);
+    public final AccompanyingCanvas paintWith(final Minter aMinter, final String aCanvasRegion,
+            final List<ContentResource> aContentList) throws ContentOutOfBoundsException,
```
[the remark this answers is not in the document]
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java:304]: This has that on line 305

File: src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (markmatney) [reply src/main/java/info/freelibrary/iiif/presentation/v3/AccompanyingCanvas.java:296]: oops, sorry

File: src/main/java/info/freelibrary/iiif/presentation/v3/PlaceholderCanvas.java [hunk 3]
```diff
@@ -239,78 +269,87 @@ public PlaceholderCanvas setWidthHeight(final int aWidth, final int aHeight) {
...
+
     }
 
     @Override
     @SafeVarargs
-    public final PlaceholderCanvas supplementWith(final MediaFragmentSelector aCanvasRegion,
-            final ContentResource... aContentArray) throws IllegalArgumentException {
-        return (PlaceholderCanvas) super.supplementWith(aCanvasRegion, aContentArray);
+    public final PlaceholderCanvas supplementWith(final Minter aMinter, final MediaFragmentSelector aCanvasRegion,
+            final ContentResource... aContentArray) {
```
[the remark this answers is not in the document]
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/PlaceholderCanvas.java:329]: I wondered about this :-)

[the remark this answers is not in the document]
Comment (markmatney) [reply src/main/java/info/freelibrary/iiif/presentation/v3/PlaceholderCanvas.java:329]: IIIF-908

**Pull request merged**